### DISCUSSION TITLE
"dynamic mixins"

### DISCUSSION THREAD

**Discussion by three-old-coders** [remkop/picocli#1664] [Ideas]
Hi! We use several static classes to group CLI params using mixins. This works well. Now I am looking for a solution to remove mixins dynamically before calling execute.
As an example: Our software runs on different environments and needs different params (some args are windows specific like a path, which is not used on a host and vice versa). To use derived classes will produce many redundant classes. So I am looking for a way to remove mixins on runtime or to intercept the addMixin method. Would be nice to have some kind of predicate callback for each created option to decide if it is used or to be ignored based on a runtime evaluation/criteria (eg Sysem.getProperty("os.name") in our case)

Thanks in advance
Jens

**Comment by remkop** [comment]
Are you aware of this feature? https://picocli.info/#_model_transformations

Sounds like that may match your needs.

**Reply by three-old-coders** [reply]
That's what I tried, too. Unfortunately I did not find a way to remove a mixin or option even the docs state:
"This allows applications to dynamically add or remove options, positional parameters or subcommands, or modify the command in any other way, based on some runtime condition"
All returned collections are "unmodifyable" so can't be manipulated directly and I do not see any method beside "removeSubcommand". Any other hint?

**Reply by remkop** [reply]
Sorry for the late reply.
True, there is no `CommandSpec::removeMixin` method.
But there is a [`CommandSpec::remove(ArgSpec)` method](https://picocli.info/apidocs-all/info.picocli/picocli/CommandLine.Model.CommandSpec.html#remove(picocli.CommandLine.Model.ArgSpec)) for removing options and positional parameters.

**Reply by three-old-coders** [reply]
Sorry, now I am a little late: 

I defined the ArgGroup as exlusive=false and my static inner class has multiple options.
How do find the ArgSpec now which is used as parameter to the remove method?

**Reply by remkop** [reply]
You could use [CommandSpec::findOption(char)](https://picocli.info/apidocs-all/info.picocli/picocli/CommandLine.Model.CommandSpec.html#findOption(char)) or [CommandSpec::findOption(String)](https://picocli.info/apidocs-all/info.picocli/picocli/CommandLine.Model.CommandSpec.html#findOption(java.lang.String)) for named options. For positional parameters I think there is only the [CommandSpec::positionalParameters](https://picocli.info/apidocs-all/info.picocli/picocli/CommandLine.Model.CommandSpec.html#positionalParameters()) method.

**Reply by remkop** [reply]
Oops, I did not realize your application uses ArgGroups.
When using ArgGroups, it may not be easy to remove individual options. 

The bookkeeping for groups is such that options that are members of ArgGroups are registered in multiple places. I don't think that the `remove(ArgSpec)` method handles this (removal of options that are part of ArgGroups) correctly.

**Comment by remkop** [comment]
Given the ArgGroups issue, perhaps a different approach is needed.

One idea is to combine the declarative (annotations) API with the dynamic (programmatic) API:

1. create the basic application (meaning, the subcommands and options that do not vary per environment) using the annotations
2. then add mixins, subcommands and/or options dynamically for different environments using the programmatic API
3. and finally call the `execute` method
